rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala

```diff
@@ -0,0 +1,273 @@
+package coop.rchain.rholang.interpreter
+
+import cats._
+import cats.data._
+import cats.implicits._
+
+import scala.annotation.tailrec
+import scala.collection.immutable.Stream
+
+object SpatialMatcher {
+  type FreeMap            = Map[Int, Par]
+  type OptionalFreeMap[A] = StateT[Option, FreeMap, A]
+  type NonDetFreeMap[A]   = StateT[Stream, FreeMap, A]
+
+  def possiblyRemove[T](needle: T, haystack: List[T]): Option[List[T]] = {
+    val (before, after) = haystack.span(x => x != needle)
+    after match {
+      case Nil       => None
+      case _ :: tail => Some(before ++ tail)
+    }
+  }
+
+  @tailrec
+  def possiblyFind[T, R](prop: (T => Option[R]), haystack: List[T]): Option[R] =
```

Comment:
`(T => Option[R])` brackets are not needed, `T => Option[R]` will suffice

Reply:
Done.